C++ API parity: Softshrink

pbelevich opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #27309 C++ API parity: MultiheadAttention
* #27382 C++ API parity: Linear
* #27538 C++ API parity: Threshold
* #27537 C++ API parity: Tanhshrink
* #27536 C++ API parity: Tanh
* #27535 C++ API parity: Softsign
* **#27534 C++ API parity: Softshrink**

Differential Revision: [D17835404](https://our.internmc.facebook.com/intern/diff/D17835404)

pbelevich committed:
C++ API parity: Softshrink

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: Softshrink"




[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: Softshrink"




[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: Softshrink"




[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: Softshrink"




[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: Softshrink"




[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: Softshrink"




[ghstack-poisoned]

yf225 commented on test/cpp/api/functional.cpp:
```diff
@@ -554,3 +554,38 @@ TEST_F(FunctionalTest, SoftplusDefaultOptions) {
+  for (const auto lambda : {0.0, 0.42, 1.0, 4.2, 42.42}) {
+    auto x = torch::linspace(-10.0, 10.0, size * size * size);
+    x.resize_({size, size, size}).set_requires_grad(true);
+    auto y = F::softshrink(x, SoftshrinkOptions(lambda));
```
We can change this to
```suggestion
    auto y = F::softshrink(x, /*lambda=*/lambda);
```
to also test `SoftshrinkOptions`'s implicit constructor :D

yf225 commented on test/cpp/api/modules.cpp:
```diff
@@ -1258,6 +1258,25 @@ TEST_F(ModulesTest, Softplus) {
+TEST_F(ModulesTest, Softshrink) {
+  const auto size = 3;
+  for (const auto lambda : {0.0, 0.42, 1.0, 4.2, 42.42}) {
+    Softshrink model {SoftshrinkOptions(lambda)};
```
We can change this to
```suggestion
    Softshrink model {/*lambda=*/lambda};
```
to also test `SoftshrinkOptions`'s implicit constructor :D

pbelevich committed:
Update on "C++ API parity: Softshrink"


Differential Revision: [D17835404](https://our.internmc.facebook.com/intern/diff/D17835404)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: Softshrink"


Differential Revision: [D17835404](https://our.internmc.facebook.com/intern/diff/D17835404)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: Softshrink"


Differential Revision: [D17835404](https://our.internmc.facebook.com/intern/diff/D17835404)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: Softshrink"


Differential Revision: [D17835404](https://our.internmc.facebook.com/intern/diff/D17835404)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: Softshrink"


Differential Revision: [D17835404](https://our.internmc.facebook.com/intern/diff/D17835404)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: Softshrink"


Differential Revision: [D17835404](https://our.internmc.facebook.com/intern/diff/D17835404)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: Softshrink"


Differential Revision: [D17835404](https://our.internmc.facebook.com/intern/diff/D17835404)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: Softshrink"


Differential Revision: [D17835404](https://our.internmc.facebook.com/intern/diff/D17835404)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: Softshrink"


Differential Revision: [D17835404](https://our.internmc.facebook.com/intern/diff/D17835404)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: Softshrink"


Differential Revision: [D17835404](https://our.internmc.facebook.com/intern/diff/D17835404)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: Softshrink"


Differential Revision: [D17835404](https://our.internmc.facebook.com/intern/diff/D17835404)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: Softshrink"


Differential Revision: [D17835404](https://our.internmc.facebook.com/intern/diff/D17835404)

[ghstack-poisoned]

yf225 reviewed (approved):
Thanks a lot for the awesome work @pbelevich!

pbelevich committed:
Update on "C++ API parity: Softshrink"


Differential Revision: [D17835404](https://our.internmc.facebook.com/intern/diff/D17835404)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: Softshrink"


Differential Revision: [D17835404](https://our.internmc.facebook.com/intern/diff/D17835404)

[ghstack-poisoned]
